src/acl/Arp.h

```diff
@@ -9,15 +9,12 @@
 #ifndef SQUID_ACLARP_H
 #define SQUID_ACLARP_H
 
+#if USE_SQUID_EUI
```

Comment:
> Francesco: Right now a default build works on my system, but I am wondering what to do about a minimal build. The problem at hand is that src/acl/Arp.h unconditionally includes eui/Eui48.h, which is empty in case USE_SQUID_EUI is off. Result is:
>
> ```
> In file included from Arp.hdrtest.cc:15:
> ../../../../src/acl/Arp.h:34:22: error: ‘Eui’ was not declared in this scope
>    34 |     typedef std::set<Eui::Eui48> AclArpData_t;
> ```
>
> What is the best way forward in your opinion? I could make an empty Eui48 class in case USE_SQUID_EUI is off, or chase the rabbit down the hole and not define the Arp ACL and delete all references to it all over the place. Thoughts?

Our "make check" should not be testing acl/Arp.h header when our "make" does not include acl/Arp.h.

Same for other conditionally used headers, of course.

Reply:
I'm fine with this, it requires an include guard as discussed [here](https://github.com/squid-cache/squid/pull/1463/files#r1330568697). If you agree with the principle there, implementation is simple

Reply:
> I'm fine with this, it requires an include guard as discussed [here](https://github.com/squid-cache/squid/pull/1463/files#r1330568697). If you agree with the principle there, implementation is simple

I failed to find the principle in that discussion, but I did propose the one I can agree with: "make check" should not be testing headers that "make" is not using (roughly speaking). I think that principle applies to this ENABLE_EUI code well. It does not require an include guard here (or there) AFAICT. It requires Makefile.am adjustments.

Reply:
Looking at the generated Makefiles, SOURCES includes EXTRA_*_SOURCES as well, and according to https://www.gnu.org/software/automake/manual/html_node/Program-and-Library-Variables.html#index-EXTRA_005fmaude_005fSOURCES we need to have these declared in here. In other words, at least here, repeating the Eui48.h header gard into acl/Arp.h seems to be the only path forward

Reply:
> Looking at the generated Makefiles, SOURCES includes EXTRA_*_SOURCES as well

Agreed.

> at least here, repeating the Eui48.h header gard into acl/Arp.h seems to be the only path forward

I disagree. AFAICT, the paths forward may include:

1. Add guards to every conditionally compiled header file.
2. Always compute the list of tested headers manually.
3. Compute the list of tested headers manually where SOURCES includes headers that should not be tested.
4. Use SOURCES but automatically skip headers that are also listed in EXTRA_DIST??
5. Use SOURCES but automatically skip manually listed headers.

Option 1 introduces an inconsistency in primary sources (conditionally used header files need conditional compilation guards but other conditionally used source files do not). It results in same guards checked in external source files and in header files. It requires extra work in every conditionally compiled header file. It violates the very natural "`make check` only what `make all` uses" principle. Right now, it is my least favorite option.

If fully-automated option 4 does not work, then option 5 is my current favorite. Since we already have to write manual rules for conditional compilation, it would not be difficult to add an extra EXCLUDED_FROM_TESTING line:

```Makefile
if ENABLE_EUI
libacls_la_SOURCES += $(ARP_ACLS)
else
EXCLUDED_FROM_TESTING += $(ARP_ACLS)
endif
EXTRA_libacls_la_SOURCES += $(ARP_ACLS)
```

... and then automatically skip sources on the EXCLUDED_FROM_HEADER_TESTING list.

N.B. With all other factors being equal, we should prefer options that risk forgetting to exclude a header from testing (e.g., option 1 and option 5) than options that risk forgetting to test a header (e.g., option 2 and option 3).

Reply:
I managed to do this; see https://github.com/squid-cache/squid/pull/1463#issuecomment-1731419501